### PR TITLE
Update South Korea holidays, add l10n support

### PR DESCRIPTION
**Pull request by PPsyrius**
## Proposed change

- Add South Korea holidays (`en_US`, **`kr`**, `th` localization).
- Implement Election Date Calculation (as per [`Article 34 of the Public Official Election Act`](https://elaw.klri.re.kr/eng_service/%20lawView.do?hseq=38405&lang=ENG) ).
- Inclusion of all [`Temporary Public Holidays`](https://namu.wiki/w/%EC%9E%84%EC%8B%9C%EA%B3%B5%ED%9C%B4%EC%9D%BC) from 1948-2023 as part of the `StaticHolidays` class.
- Revise holiday logic work from the [1st-18th amendment of the `Public Holidays Act`](https://namu.wiki/w/%EA%B3%B5%ED%9C%B4%EC%9D%BC/%EB%8C%80%ED%95%9C%EB%AF%BC%EA%B5%AD/%EC%97%AD%EC%82%AC).

## Type of change

- [ ] New country/market holidays support (thank you!)
- [x] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've added references to all holidays information sources used in this PR
- [x] The code style looks good: `make pre-commit` command generates no changes
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/beta/docs/source

### CODE REVIEW

**Comment by PPsyrius** [comment]
This PR is marked as a draft to wait for #1535 's eventual merge (mostly so that we can include the historic, short observances in 1959-1960, 1989 done (as seen [here](https://namu.wiki/w/%EB%8C%80%EC%B2%B4%20%ED%9C%B4%EC%9D%BC%20%EC%A0%9C%EB%8F%84); Point `4.2.1 
Before the 2010s` ).

**Review by KJhellico — COMMENTED** [review]
@PPsyrius, it's just great!

File: holidays/countries/south_korea.py [hunk 1]
```diff
@@ -90,75 +119,174 @@ def _add_alt_holiday(
...
+                    # Independence Movement Day (MAR, 1).
+                    self._add_holiday_2nd_wed_of_mar(name)
+                else:
+                    self._add_holiday_1st_wed_of_mar(name)
+
+        # Local Election Day.
+        name = tr("지방선거일")
+
+        if self._year >= 2007 and (self._year - 2010) % 4 == 0:
+            jun_6 = (JUN, 6)
```
[the remark this answers is not in the document]
Comment (PPsyrius) [reply holidays/countries/south_korea.py:276]: I thought that wouldn't work at first, but it surprisingly did huh :eyes:

Fixed now <3

**Comment by PPsyrius** [comment]
Turns out #1535's new `special_holidays_observed` requires the usage of `ObservedHolidayBase`

We may as well go and revamp the existing codes for `_add_alt_holiday( )` and `_add_three_day_holiday( )` at this rate, as the change to the new observed class requires at least one `observed_rule`.

**Comment by KJhellico** [comment]
I'm trying to migrate South Korea to ObservedHolidays, but I'm not satisfied with the result yet. Their rules are quite complex and unique.

File: holidays/countries/south_korea.py [hunk 2]
```diff
@@ -90,75 +122,172 @@ def _add_alt_holiday(
...
+                # Folk Day.
+                else tr("민속의 날")
+            )
+            korean_new_year = self._add_chinese_new_years_day(name)
+            if self._year >= 1989:
+                self._add_three_day_holiday(korean_new_year, name)
 
         # Independence Movement Day.
-        self._add_alt_holiday(self._add_holiday_mar_1("Independence Movement Day"), since=2022)
+        mar_1 = self._add_holiday_mar_1(tr("삼일절"))
```
Comment (arkid15r) [inline holidays/countries/south_korea.py:173]:
```suggestion
        mar_1 = self._add_holiday_mar_1(tr("삼일절"))
        # mar_1 is used later for Presidential Election Day.
```

File: holidays/countries/south_korea.py [hunk 3]
```diff
@@ -90,75 +122,172 @@ def _add_alt_holiday(
...
         if self._year >= 1975:
-            self._add_alt_holiday(self._add_holiday_may_5("Children's Day"), since=2015)
+            # Children's Day.
+            self._add_alt_holiday(self._add_holiday_may_5(tr("어린이날")), since=2015)
 
-        # Memorial Day.
-        self._add_holiday_jun_6("Memorial Day")
+        if self._year >= 1956:
+            # Memorial Day.
+            jun_6 = self._add_holiday_jun_6(tr("현충일"))
```
Comment (arkid15r) [inline holidays/countries/south_korea.py:196]:
```suggestion
            jun_6 = self._add_holiday_jun_6(tr("현충일"))
            # jun_6 is  used later for Local Election Day.
```

File: holidays/countries/south_korea.py [hunk 4]
```diff
@@ -90,75 +122,172 @@ def _add_alt_holiday(
...
+
+        if self._year == 2020:
+            self._add_holiday_apr_15(name)
+        elif self._year >= 2007 and (self._year - 2008) % 4 == 0:
+            self._add_holiday_2nd_wed_of_apr(name)
+
+        # Presidential Election Day.
+        name = tr("대통령 선거일")
+
+        if self._year >= 2007 and (self._year - 2007) % 5 == 0:
```
Comment (arkid15r) [inline holidays/countries/south_korea.py:259]:
```suggestion
        if self._year >= 2007 and (self._year - 2007) % 5 == 0:
            # Presidential Election Day.
            name = tr("대통령 선거일")

```

File: holidays/countries/south_korea.py [hunk 5]
```diff
@@ -90,75 +122,172 @@ def _add_alt_holiday(
...
+                    # Moved as per Paragraph 2 of Article 34 due to conflict with
+                    # Independence Movement Day (MAR, 1).
+                    self._add_holiday_2nd_wed_of_mar(name)
+                else:
+                    self._add_holiday_1st_wed_of_mar(name)
+
+        # Local Election Day.
+        name = tr("지방선거일")
+
+        if self._year >= 2007 and (self._year - 2010) % 4 == 0:
```
Comment (arkid15r) [inline holidays/countries/south_korea.py:277]:
```suggestion
        if self._year >= 2007 and (self._year - 2010) % 4 == 0:
            # Local Election Day.
            name = tr("지방선거일")

```

**Review by arkid15r — COMMENTED** [review]
Smooth migration plus l10n, 2 in 1!

A few suggestions:

**Review by arkid15r — APPROVED** [review]
🇰🇷 LGTM

**Pull request merged**